src/xrpld/app/misc/detail/LendingHelpers.cpp

```diff
@@ -0,0 +1,2072 @@
+    TenthBips32 const closeInterestRate,
+    std::int32_t loanScale,
+    Number const& totalInterestOutstanding,
+    Number const& periodicRate,
+    Number const& closePaymentFee,
+    STAmount const& amount,
+    TenthBips16 managementFeeRate,
+    beast::Journal j)
+{
+    // Full payment must be made before the final scheduled payment.
+    if (paymentRemaining <= 1)
+    {
+        // If this is the last payment, it has to be a regular payment
+        JLOG(j.warn()) << "Last payment cannot be a full payment.";
+        return Unexpected(tecKILLED);
+    }
+
+    // Calculate the theoretical principal based on the payment schedule.
+    // This raw (unrounded) value is used to compute interest and penalties
+    // accurately.
+    Number const rawPrincipalOutstanding = loanPrincipalFromPeriodicPayment(
+        periodicPayment, periodicRate, paymentRemaining);
+
+    // Full payment interest includes both accrued interest (time since last
+    // payment) and prepayment penalty (for closing early).
+    auto const fullPaymentInterest = computeFullPaymentInterest(
+        rawPrincipalOutstanding,
+        periodicRate,
+        view.parentCloseTime(),
+        paymentInterval,
+        prevPaymentDate,
+        startDate,
+        closeInterestRate);
+
+    // Split the full payment interest into net interest (to vault) and
+    // management fee (to broker), applying proper rounding.
+    auto const [roundedFullInterest, roundedFullManagementFee] = [&]() {
+        auto const interest = roundToAsset(
+            asset, fullPaymentInterest, loanScale, Number::downward);
+        auto const parts = computeInterestAndFeeParts(
+            asset, interest, managementFeeRate, loanScale);
+        return std::make_tuple(parts.first, parts.second);
+    }();
+
+    ExtendedPaymentComponents const full{
+        PaymentComponents{
+            // Pay off all tracked outstanding balances: principal, interest,
+            // and fees.
+            // This marks the loan as complete (final payment).
+            .trackedValueDelta = principalOutstanding +
+                totalInterestOutstanding + managementFeeOutstanding,
+            .trackedPrincipalDelta = principalOutstanding,
+
+            // All outstanding management fees are paid. This zeroes out the
+            // tracked fee balance.
+            .trackedManagementFeeDelta = managementFeeOutstanding,
+            .specialCase = PaymentSpecialCase::final,
+        },
+
+        // Untracked management fee includes:
+        // 1. Close payment fee (fixed fee for early closure)
+        // 2. Management fee on the full payment interest
+        // 3. Minus the outstanding tracked fee (already accounted for above)
+        // This can be negative because the outstanding fee is subtracted, but
+        // it gets combined with trackedManagementFeeDelta in the final
+        // accounting.
+        closePaymentFee + roundedFullManagementFee - managementFeeOutstanding,
```

Comment:
@ximinez is `roundedFullManagementFee` simply the same as `managementFeeOutstanding`  except it may have different rounding since it was computed? Or does it include other components?

Reply:
> @ximinez is `roundedFullManagementFee` simply the same as `managementFeeOutstanding` except it may have different rounding since it was computed? Or does it include other components?

`roundedFullManagementFee` is the management fee part of the computed `fullPaymentInterest`. That is a completely separate computation than the amount owed in the normal amortization schedule, and could be drastically different than `totalInterestOutstanding`. Thus `roundedFullManagementFee` could be drastically different than `managementFeeOutstanding`.